scarlet/source.py

```diff
@@ -72,159 +213,168 @@ def __init__(self, model_frame, sky_coord, observations):
-            constraint=PositivityConstraint(),
+        center = np.array(model_frame.get_pixel(sky_coord), dtype="float")
+        morphology = PointSourceMorphology(center, model_frame.psf)
+
+        super().__init__(model_frame, spectrum, morphology)
+
+
+class StarletMorphology(Morphology):
+    def __init__(self, image, bbox=None, threshold=0):
+
+        if bbox is None:
+            self._bbox = Box(image.shape)
+        else:
+            assert bbox.shape == image.shape
+            self._bbox = bbox
+
+        # Starlet transform of morphologies (n1,n2) with 4 dimensions: (1,lvl,n1,n2), lvl = wavelet scales
+        self.transform = Starlet(image)
+        # The starlet transform is the model
+        coeffs = self.transform.coefficients
+        # wavelet-scale norm
+        starlet_norm = self.transform.norm
+        # One threshold per wavelet scale: thresh*norm
+        thresh_array = np.zeros(coeffs.shape) + threshold
+        thresh_array = (
+            thresh_array * np.array([starlet_norm])[..., np.newaxis, np.newaxis]
         )
-        center = Parameter(self.center, name="center", step=1e-1)
+        # We don't threshold the last scale
+        thresh_array[:, -1, :, :] = 0
 
-        # define bbox
-        pixel_center = tuple(np.round(center).astype("int"))
-        front, back = 0, C
-        bottom = pixel_center[0] - model_frame.psf.shape[1] // 2
-        top = pixel_center[0] + model_frame.psf.shape[1] // 2
-        left = pixel_center[1] - model_frame.psf.shape[2] // 2
-        right = pixel_center[1] + model_frame.psf.shape[2] // 2
-        bbox = Box.from_bounds((front, back), (bottom, top), (left, right))
+        constraint = ConstraintChain(L0Constraint(thresh_array), PositivityConstraint())
 
-        super().__init__(model_frame, bbox, sed, center, self._psf_wrapper)
+        coeffs = Parameter(coeffs, name="coeffs", step=1e-2, constraint=constraint)
+        super().__init__(coeffs)
+
+    def get_model(self, *parameters):
+        """ Takes the inverse transform of parameters as starlet coefficients.
+
+        """
+        coeffs = self.parameters[0]
+        for p in parameters:
+            if p._value.name == "coeffs":
+                coeffs = p
+        return Starlet(coefficients=coeffs).image[0]
 
-    def _psf_wrapper(self, *parameters):
-        return self.model_frame.psf.__call__(*parameters, bbox=self.bbox)[0]
 
-class StarletSource(FunctionComponent):
+class StarletSource(FactorizedComponent):
     """Source intialized with starlet coefficients.
 
     Sources are initialized with the SED of the center pixel,
     and the morphologies are initialised as ExtendedSources
     and transformed into starlet coefficients.
     """
+
     def __init__(
         self,
-        frame,
+        model_frame,
         sky_coord,
         observations,
         coadd=None,
         bg_cutoff=None,
         thresh=1.0,
-        starlet_thresh=5,
         min_grad=0.1,
+        starlet_thresh=5,
```

Comment:
I'd like to be able to give an input sed, and also to allow the box to be the entire image with an optional parameter. I can do that in an upcoming pr. Also, being able to limit the step size of the sed parameter would be great.

Reply:
See above. Happy to do that.

As for the spectrum (!) step size, this should be set in `StarletMorphology` where the parameter is created.